core/Extensible.php

```diff
@@ -0,0 +1,465 @@
+	}
+
+	/**
+	 * Allows user code to hook into Object::extend after control
+	 * being delegated to extensions. Each callback will be reset
+	 * once called.
+	 *
+	 * @param string $method The name of the method to hook into
+	 * @param callable $callback The callback to execute
+	 */
+	protected function afterExtending($method, $callback) {
+		if(empty($this->afterExtendCallbacks[$method])) {
+			$this->afterExtendCallbacks[$method] = array();
+		}
+		$this->afterExtendCallbacks[$method][] = $callback;
+	}
+
+	protected function constructExtensions() {
+		$class = get_class($this);
+
+		// Register this trait as a method source
+		$this->registerExtraMethodCallback('defineExtensionMethods', function() {
+			$this->defineExtensionMethods();
+		});
+
+		// Setup all extension instances for this instance
+		foreach(ClassInfo::ancestry($class) as $class) {
+			if(in_array($class, self::$unextendable_classes)) continue;
+			$extensions = Config::inst()->get($class, 'extensions',
+				Config::UNINHERITED | Config::EXCLUDE_EXTRA_SOURCES);
+
+			if($extensions) foreach($extensions as $extension) {
+				$instance = \Object::create_from_string($extension);
+				$instance->setOwner(null, $class);
+				$this->extension_instances[$instance->class] = $instance;
+			}
+		}
+
+		if(!isset(self::$classes_constructed[$class])) {
+			$this->defineMethods();
+			self::$classes_constructed[$class] = true;
+		}
+	}
+
+	/**
+	 * Adds any methods from {@link Extension} instances attached to this object.
+	 * All these methods can then be called directly on the instance (transparently
+	 * mapped through {@link __call()}), or called explicitly through {@link extend()}.
+	 *
+	 * @uses addMethodsFrom()
+	 */
+	protected function defineExtensionMethods() {
+		if(!empty($this->extension_instances)) {
+			foreach (array_keys($this->extension_instances) as $key) {
+				$this->addMethodsFrom('extension_instances', $key);
+			}
+		}
+	}
+
+
+	/**
+	 * Add an extension to a specific class.
+	 *
+	 * The preferred method for adding extensions is through YAML config,
+	 * since it avoids autoloading the class, and is easier to override in
+	 * more specific configurations.
+	 *
+	 * As an alternative, extensions can be added to a specific class
+	 * directly in the {@link Object::$extensions} array.
+	 * See {@link SiteTree::$extensions} for examples.
+	 * Keep in mind that the extension will only be applied to new
+	 * instances, not existing ones (including all instances created through {@link singleton()}).
+	 *
+	 * @see http://doc.silverstripe.org/framework/en/trunk/reference/dataextension
+	 * @param string $classOrExtension Class that should be extended - has to be a subclass of {@link Object}
+	 * @param string $extension Subclass of {@link Extension} with optional parameters
+	 *  as a string, e.g. "Versioned" or "Translatable('Param')"
+	 * @return bool Flag if the extension was added
+	 */
+	public static function add_extension($classOrExtension, $extension = null) {
```

Comment:
Config updates can't live-update extensions; It requires an event to re-define methods, etc, which tests rely on.